### PR TITLE
[next] support middleware-manifest v2

### PR DESCRIPTION
**Pull request by nkzawa**
### Related Issues

As the title, support a new version of middleware-manifest of next.js that is going to be added in https://github.com/vercel/next.js/pull/39257

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

**Review by ijjk — COMMENTED** [review]
I'm seeing the below error while testing, we may want to clean up the error as it's not exactly clear which route was incorrect and the expected values. 

```sh
21:55:12.935 | Created all serverless functions in: 46.001ms
21:55:12.967 | Collected static files (public/, static/, .next/static): 2.554ms
21:55:13.161 | Build Completed in /vercel/output [24s]
21:55:13.646 | Error: Mapping middleware not found. Maybe you provided a wrong middlewarePath? Available paths: middleware
```

File: packages/next/src/utils.ts [hunk 1]
```diff
@@ -2157,23 +2163,44 @@ export {
...
+  middleware: { [page: string]: EdgeFunctionInfoV1 };
+  functions?: { [page: string]: EdgeFunctionInfoV1 };
+}
+
+interface MiddlewareManifestV2 {
+  version: 2;
+  sortedMiddleware: string[];
+  middleware: { [page: string]: EdgeFunctionInfoV2 };
+  functions?: { [page: string]: EdgeFunctionInfoV2 };
 }
```
Comment (feugy) [inline packages/next/src/utils.ts:2180]: **thought:** what a shame typescript does not allow us extending interfaces and overriding some of the fields

File: packages/next/src/utils.ts [hunk 2]
```diff
@@ -2390,36 +2421,74 @@ export async function getMiddlewareManifest(
     return;
   }
 
-  return fs.readJSON(middlewareManifestPath);
+  const manifest = (await fs.readJSON(
+    middlewareManifestPath
+  )) as MiddlewareManifest;
+  return manifest.version === 1
+    ? upgradeMiddlewareManifest(manifest)
```
Comment (feugy) [inline packages/next/src/utils.ts:2431]: **praise:** smart! I was about to ask how `getRouteMatchers` could be backward compatible. It does not have to be! 👏🏻

**Review by feugy — APPROVED** [review]
Besides JJ's request for better error handling, no extra requests.

File: packages/next/src/utils.ts (hunk 1 above)
Comment (nkzawa) [reply packages/next/src/utils.ts:2180]: yeah, we can do that if we use `Omit<>` to exclude overriding fields but not sure if it's better in this case.

**Comment by nkzawa** [comment]
The error @ijjk mentioned should have already been fixed by https://github.com/vercel/api/pull/13624

**Review by ijjk — COMMENTED** [review]
The changes here look good, will hold off on approving until the e2e deploy tests are passing on the Next.js PR though

**Review by ijjk — APPROVED** [review]
Thanks!

**Pull request merged**